src/browser/transports/injection_bus.ts

```diff
@@ -88,28 +91,84 @@ export default class NativeWindowInjectionBus extends EventEmitter {
+
       if (this._pendingRequests.has(messageId)) {
-        const isNack = parsedMessage.action.includes('error');
-        const pendingRequest = this._pendingRequests.get(messageId);
+        this._listenerAck(parsedMessage);
+      } else {
+        this._listenerBroadcastMsg(parsedMessage);
+      }
+    };
 
-        isNack
-          ? pendingRequest.reject((<NackMessage>parsedMessage).payload.reason)
-          : pendingRequest.resolve();
+    copyDataTransport.on('message', this._messageListener);
+
+    this.subscribe();
+  }
 
-        this._senderId = senderId;
-        this._pendingRequests.delete(messageId);
+  // Setup heartbeat
+  private setupHeartbeat() {
+    this._connected = false;
+    this._previousDllSequence = -1;
 
-        return;
+    this._listenerHeartbeat = (message: MessageBase): void => {
+      const { sequence } = message;
+
+      if (this._previousDllSequence + 1 !== sequence) {
+        // At least one message has been missed
+        const msg = `[NWI] Missed message(s) from ${this._senderId}: `
+          + `previous sequence ${this._previousDllSequence}, `
+          + `current sequence: ${sequence}`;
+        writeToLog('info', msg);
+      }
+
+      this._previousDllSequence = sequence;
+    };
+
+    // Ping message window every second and update current connection status
+    setInterval(() => {
+      this.send({
+        action: 'status/ping/request',
+        payload: {}
+      }).then(() => {
+        if (!this._connected) {
+          // This indicates that the core re-connected,
+          // so we need to subscribe to events again
+          this.subscribe();
+        }
+        this._connected = true;
+      }).catch(() => {
+        this._connected = false;
+      });
+    }, 1000);
+  }
+
+  // Setup listener responsible for parsing acks
+  private setupListenerAck() {
+    this._listenerAck = (message: MessageBase): void => {
+      const { messageId, senderId } = message;
+      const isNack = message.action.includes('error');
+      const pendingRequest = this._pendingRequests.get(messageId);
+
+      if (isNack) {
+        pendingRequest.reject((<NackMessage>message).payload.reason);
+      } else {
+        pendingRequest.resolve();
       }
 
-      // Broadcast message
-      if (!(<BroadcastMessage>parsedMessage).payload || !(<BroadcastMessage>parsedMessage).payload.data) {
-        writeToLog('info', `[NWI] Injection event without payload: ${JSON.stringify(parsedMessage)}`); // TODO
+      this._senderId = senderId;
+      this._pendingRequests.delete(messageId);
+    };
+  }
+
+  // Setup listener responsible for parsing broadcast messages
+  private setupListenerBroadcastMsg() {
+    this._listenerBroadcastMsg = (message: MessageBase): void => {
+      if (!(<BroadcastMessage>message).payload || !(<BroadcastMessage>message).payload.data) {
+        writeToLog('info', `[NWI] Injection event without payload: ${JSON.stringify(message)}`);
```

Comment:
Is this actionable? Should we make an event out of this?

Reply:
This was added initially and left there while we are ironing out parts of the communication protocol with the injection. I would like to  eventually remove it, this is here for now to make sure it doesn't throw.